[<h1> FIT5197 Assignment 3 Semester 2, 2020 </h1>](https://lms.monash.edu/mod/assign/view.php?id=7560449)

---
Authors: Dan Nguyen, Yun Zhao

Admins (Competition): Dr. Levin Kuhlmann, Yun Zhao, Anil Gurbuz

Proofreaders: Dr. Levin Kuhlmann, Yun Zhao, and other tutors 

Date: Oct 2020

---

[<h1> Assignment Instruction </h1>](https://lms.monash.edu/mod/assign/view.php?id=7560449)

<span style="color:red"> Please read through the instructions carefully, by submitting the assignment, you are considered to have read all the instructions carefully and be aware of the penalties that entail. </span>

<h1>Part 1: Regression (50 Marks)</h1>

This part is about regression. Specifically, you will be ``predicting the fuel efficiency`` of a car (in kilometers per litre) based on its characteristics. This is a practical problem as Australia is one of the largest automobile markets in the world; thus, correctly predicting the fuel efficiency is necessary to control emission rates to the environment.

The dataset has many observations and predictors obtained from many retailers for car models available for sale from 2017 to 2020. The target variable is the fuel efficiency of the car measured in kilometers per litre. The higher this value, the better the fuel efficiency of the car. 

PleaseProvide working/R code/justifications for each of these questions as required.

$\textbf{Note:}$ If not explicitly mentioned, libraries are not allowed

In [1]:
# Read the data from students' side
remove(list = ls())
train <- read.csv("RegressionTrain.csv")
test <- read.csv("RegressionTest.csv")

In [ ]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# Please skip (don't run) this if you are a student
# Read in the data from marking tutors' side (ensure no cheating!)
remove(list = ls())
train <- read.csv("../data/RegressionTrain.csv")
test <- read.csv("../data/RegressionTest.csv")
label <- read.csv("../data/RegressionTestLabel.csv")

<h2> Question 1 (5 Marks) </h2>

Fit a $\textbf{multiple linear model}$ to the fuel efficiency data using the ``train`` dataset. By checking the summary information, which predictors/variables do you think are possibly associated with fuel efficiency (use ``0.05`` significant level), and why? Which ``three predictors/variables`` appear to be the strongest predictors of fuel efficiency, and why? 

$\textbf{Note}$: You don't have to worry about categorical variables here since R can deal with this automatically, focus your efforts on interpretation. Additionally, when explaining why features are strongly associated with the target, please refrain giving one or two sentences answers, these answers are not descriptive enough and will result in deduction of marks. Finally, please name the model here ``lm.fit`` for future marking purposes.

$\textbf{YOUR ANSWER HERE}$  
When we use the summary information, we find that Model.Year, Eng.Displacement, No.Cylinders, Aspiration-SC， Aspiration-TC, Aspiration-TS, No.Gears, Lockup.Torque.Converter-Y, Drive.Sys-F, Fuel.Type-GP, and Max.Ethanol are possibly associated with fuel efficiency because from the summary result, we know that they satisfied the 0.05 significant level.  
Then, we find that the values with the smallest significant level are Aspiration-SC, Aspiration-TC, Lockup.Torque.Converter-Y, Drive.Sys-F, Eng.Displacement, No.Cylinders, and No.Gears ( *** in summary result, means the significant level approximate to 0).   
Moreover, we find that Eng.Displacement, No.Cylinders, and No.Gears variables are continuous variables and seem to be (normal-ish) distributed. Also, fuel efficiency values also look like normal distribution.  
In summary, we can make a conclusion that Eng.Displacement, No.Cylinders, and No.Gears variables appear to be the strongest predictors of fuel efficiency.


In [2]:
# Change this
lm.fit <-  lm(Comb.FE ~ ., train)
summary(lm.fit)


Call:
lm(formula = Comb.FE ~ ., data = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.0256 -0.9978 -0.0644  0.7006 11.3941 

Coefficients:
                           Estimate Std. Error t value Pr(>|t|)    
(Intercept)              -1.783e+02  8.587e+01  -2.076  0.03809 *  
Model.Year                9.640e-02  4.255e-02   2.266  0.02363 *  
Eng.Displacement         -1.364e+00  1.025e-01 -13.306  < 2e-16 ***
No.Cylinders              4.644e-02  6.769e-02   0.686  0.49282    
AspirationOT             -3.452e-01  6.352e-01  -0.543  0.58693    
AspirationSC             -9.197e-01  2.282e-01  -4.031 5.85e-05 ***
AspirationTC             -1.303e+00  1.288e-01 -10.111  < 2e-16 ***
AspirationTS             -1.149e+00  4.945e-01  -2.323  0.02035 *  
No.Gears                 -1.307e-01  2.995e-02  -4.364 1.37e-05 ***
Lockup.Torque.ConverterY -8.243e-01  1.117e-01  -7.377 2.78e-13 ***
Drive.SysA               -8.339e-02  1.521e-01  -0.548  0.58356    
Drive.SysF                1.

<h2> Question 2 (5 Marks) </h2>

Describe/discuss the effect that the year of manufacture ``(Model.Year)`` variable appears to have on the mean ``fuel efficiency``. Additionally, describe/discuss the effect that the number of gears ``(No.Gears)`` variable has on the mean ``fuel efficiency`` of the car.

$\textbf{Note}$: This asks for your descriptions, please refrain from using one or two lines to describe/discuss the effect. Keep answers to be 4 decimal places

In [3]:
for (i in sort(unique(train$Model.Year))){
mean_value = round(mean(train$Comb.FE[which(train$Model.Year == i)] ),4)
print(paste("In the year of ", i ,", the mean of fuel efficiency is", mean_value))
    }
print("-----------------------------------------------------------------------------")
for (i in sort(unique(train$No.Cylinders))){
mean_value = round(mean(train$Comb.FE[which(train$No.Cylinders == i)] ),4)
print(paste("When the number of gears is", i ,", the mean of fuel efficiency is", mean_value))
    }

[1] "In the year of  2017 , the mean of fuel efficiency is 10.3663"
[1] "In the year of  2018 , the mean of fuel efficiency is 10.5806"
[1] "In the year of  2019 , the mean of fuel efficiency is 10.3784"
[1] "In the year of  2020 , the mean of fuel efficiency is 10.4758"
[1] "-----------------------------------------------------------------------------"
[1] "When the number of gears is 3 , the mean of fuel efficiency is 13.7678"
[1] "When the number of gears is 4 , the mean of fuel efficiency is 12.465"
[1] "When the number of gears is 5 , the mean of fuel efficiency is 9.9472"
[1] "When the number of gears is 6 , the mean of fuel efficiency is 9.5233"
[1] "When the number of gears is 8 , the mean of fuel efficiency is 7.8253"
[1] "When the number of gears is 10 , the mean of fuel efficiency is 7.1439"
[1] "When the number of gears is 12 , the mean of fuel efficiency is 6.6503"
[1] "When the number of gears is 16 , the mean of fuel efficiency is 4.9736"


$\textbf{YOUR ANSWER HERE}$  
In the different years of manufacture, the mean of fuel efficiency looks similar, the values range of it is from 10.3663 to 10.5806. So, we think that the year of manufacture has a slight impact on the mean fuel efficiency.  
On the other hand, the number of gears has a significant effect on the mean fuel efficiency.   
To be specific, the higher number of gears, the lower number of the mean fuel efficiency. When the number of gears is 3, we record the highest fuel efficiency, which is 13.7678. However, when the number of gears is 16, the mean fuel efficiency reaches the lowest point, which is 4.9736.  
On the other hand, from the linear model, we know that the mean fuel efficiency increase by 0.0964 per year, and with each additional number of gears, the mean fuel efficiency drops 0.1307.


<h2> Question 3 (5 Marks) </h2>

Apply the stepwise selection procedure with the $\textbf{BIC}$ penalty to prune out potentially less significant variables. Write down the final regression equation obtained after pruning, please keep the values of the parameter coefficients to 2 decimal places. Finally, also describe the pruned model.

$\textbf{Note}$: please don't change the default direction ``both`` in the step function, this is so that we can check your work easily. Additionally, please name this model ``sw.fit``

$\textbf{YOUR ANSWER HERE}$  
##### Introduce to BIC:
When training the model, increasing the number of parameters, namely increasing the complexity of the model, and will increase the precision of the likelihood function, but will also lead to the overfitting phenomenon. To solve this problem, AIC and BIC both introduced a penalty term related to the number of model parameters, and BIC's penalty term is larger than AIC. Considering the number of samples, large samples can effectively prevent the model complexity caused by too high precision.  
The smaller the BIC values, the better the regression equation it is.
##### The final regression equation is:
$ Comb.FE = 16.20 -1.28 \times Eng.Displacement - 0.10 \times AspirationOT -0.70 \times AspirationSC - 1.14 \times AspirationTC - 1.12 \times AspirationTS -0.11 \times No.Gears -0.82 \times No.Gears Lockup.Torque.ConverterY + 0.03 \times Drive.SysA + 1.48 \times Drive.SysF -0.32 \times Drive.SysP +  0.09 \times Drive.SysR -  0.01 \times Max.Ethanol $  
##### Describe the pruned model  
The intercept values of the pruned model is 16.20.  
From this model, we can conclude that Drive.Sys-F has a strongly positive impact on fuel efficiency, which can increase by 1.48. Also, Drive.SysA and Drive.SysR can increase fuel efficiency slightly.   
Other values will harm the fuel efficiency.   
To be specific, with 1 increase in engine displacement, the fuel efficiency will decrease 1.28. Also, AspirationTC and AspirationTS will decrease fuel efficiency by 1.14 and 1.12 respectively. The remaining variables will drop fuel efficiency mildly.


In [4]:
# Change this
sw.fit <- step(lm.fit, trace=0, k = log(nrow(train)), direction="both")

<h2> Question 4 (5 Marks) </h2>

Say we are going to buy a new car and we want to improve the fuel efficiency of our new car, what does this ``BIC model`` suggest we should do? Provide a detailed answers of at least ``150 words``.

$\textbf{YOUR ANSWER HERE}$  
From the BOC model, we know the coefficients of each variables in this multiple linear model. The function is :   
Fuel.efficiency = $16.20 -1.28 \times Eng.Displacement - 0.10 \times AspirationOT -0.70 \times AspirationSC - 1.14 \times AspirationTC - 1.12 \times AspirationTS -0.11 \times No.Gears -0.82 \times No.Gears Lockup.Torque.ConverterY + 0.03 \times Drive.SysA + 1.48 \times Drive.SysF -0.32 \times Drive.SysP +  0.09 \times Drive.SysR -  0.01 \times Max.Ethanol $   

From this function, in order to improve the fuel efficiency of our new car, we need to choose the lowest engine displacement, the number of cylinders and the number of gears. If these variables are the minimum, the fuel efficiency of this car is pretty high.  
Also, choosing F as the Drive system is a good choice, which can increase 1.48 of fuel efficiency.   
It is difficult for a car to get the minimum value of these variables at the same time, we need to make a balance between them.   
To be specific, engine displacement is the most important variable, we need to consider the lowest engine displacement at first. After that, we will consider the number of gears and the number of cylinders, also we need to choose the Drive system F.



<h2> Question 5 (5 Marks)</h2>

Imagine that you are looking for a new car to buy to replace your existing car. Use the $\textbf{test}$ dataset to inspect the first car fuel efficiency and see whether it is a good fit for you or not.
    
    (a) Use your BIC model to predict the mean fuel efficiency for this new car. Provide a 95% confidence interval for this prediction. [2 mark]
    (b) Following the previous estimation, given that the current car that you own has a mean fuel efficiency of 9.5 km/l (measured over the life time of your ownership), does your model (BIC) suggest that the new car will have better fuel efficiency than your current car? Why? [3 marks]

In [5]:
predict(sw.fit, newdata = test[1,], interval = "confidence")

fit      lwr      upr     
1 9.287257 9.052956 9.521557

$\textbf{YOUR ANSWER HERE}$  
(1) From the predict code result, we calculate a 95% confidence interval for the first car in test dataset is
$[9.0530,9.5216]$

(2)The 95% confidence interval for this prediction is $[9.0530,9.5216]$, and 9.5 km/l is near the largest value of this interval. To be specific, the possibility of this new car has better fuel efficiency is bigger than 50%, which may be 80% - 95%.   
As a result, my model (BIC) suggest that the new car will have better fuel efficiency than my current car.


<h2> Question 6 (Libraries are allowed) (25 Marks) </h2>

As a Data Scientist, one of the key tasks is to build models $\textbf{most appropriate/closest}$ to the truth; thus, modelling will not be limitted to these steps in the assignment. To simulate for a realistic modelling process, this question will be in the form of a competition among students to find out who has the best model.

Thus, You will be graded by the performance of your model compared to your classmates', the better your model, the higher your score. Additionally, you need to write a short paragraph describing/documenting your thought process in this model building process ``(300 words)``. Note that this is to explain to us why you build your current model so that we can verify that you understand the model you build and not just copy from other people.

$\textbf{Note}$ Please make sure that we can install the libraries that you use in this part, the code structure can be:

``install.packages("some package", repos='http://cran.us.r-project.org')``

``library("some package")``

Remember that if we cannot run your code, we will have to give you 0 marks, our suggestion is for you to use the standard ``R version 3.6.1``

You also need to name your final model ``fin.mod`` so we can run a check to find out your performance. A good test for your understanding would be to set the previous $\textbf{BIC model}$ to be the final model to check if your code works Appropriately.

$20$ Marks for the model performance in the competition

$5$ Marks for logically writing down the thought process in building the final model

This is the [link](https://www.kaggle.com/t/0a3c0fc91b074816a6315bb4e9b42602) to the competition

$\textbf{YOUR ANSWER HERE}$  
I write a short paragraph describing/documenting my thought process in this model building process under each code block. 

In [6]:
# Print out the relationship between fuel effeciency and each variables

# print Fuel.Type Column
for (i in sort(unique(train$Fuel.Type))){
mean_value = round(mean(train$Comb.FE[which(train$Fuel.Type == i)] ),4)
print(paste("When the fuel type is ", i ,", the mean of fuel efficiency is", mean_value))
    }
print("-----------------------------------------------------------------------------------")

# print Aspiration Column
for (i in sort(unique(train$Aspiration))){
mean_value = round(mean(train$Comb.FE[which(train$Aspiration == i)] ),4)
print(paste("When the Aspiration is ", i ,", the mean of fuel efficiency is", mean_value))
    }
print("-----------------------------------------------------------------------------------")

# print Lockup.Torque.Converter Column
for (i in sort(unique(train$Lockup.Torque.Converter))){
mean_value = round(mean(train$Comb.FE[which(train$Lockup.Torque.Converter == i)] ),4)
print(paste("When the Lockup Torque Converter is", i ,", the mean of fuel efficiency is", mean_value))
    }
print("-----------------------------------------------------------------------------------")

# print Max.Ethanol Column
for (i in sort(unique(train$Max.Ethanol))){
mean_value = round(mean(train$Comb.FE[which(train$Max.Ethanol == i)] ),4)
print(paste("When the Max.Ethanol is", i ,", the mean of fuel efficiency is", mean_value))
    }
print("-----------------------------------------------------------------------------------")

# print Drive.Sys Column
for (i in sort(unique(train$Drive.Sys))){
mean_value = round(mean(train$Comb.FE[which(train$Drive.Sys == i)] ),4)
print(paste("When the drive system is ", i ,", the mean of fuel efficiency is", mean_value))
    }

[1] "When the fuel type is  G , the mean of fuel efficiency is 11.2888"
[1] "When the fuel type is  GM , the mean of fuel efficiency is 8.2516"
[1] "When the fuel type is  GP , the mean of fuel efficiency is 10.0892"
[1] "When the fuel type is  GPR , the mean of fuel efficiency is 9.458"
[1] "-----------------------------------------------------------------------------------"
[1] "When the Aspiration is  N , the mean of fuel efficiency is 10.6972"
[1] "When the Aspiration is  OT , the mean of fuel efficiency is 10.3347"
[1] "When the Aspiration is  SC , the mean of fuel efficiency is 8.4332"
[1] "When the Aspiration is  TC , the mean of fuel efficiency is 10.4079"
[1] "When the Aspiration is  TS , the mean of fuel efficiency is 10.7384"
[1] "-----------------------------------------------------------------------------------"
[1] "When the Lockup Torque Converter is N , the mean of fuel efficiency is 11.8128"
[1] "When the Lockup Torque Converter is Y , the mean of fuel efficiency is 10

Firstly, I print out the mean of the fuel efficiency of each variable, to find out the relationship between each variable and fuel efficiency.   
This step can help us have a general knowledge of each variable, which has a positive impact on the model set.  
For example, there are five driver systems in the data. When the driver system is F, the fuel efficiency is significantly different from other driver systems.  
As a result, maybe we can divide the driver systems into two groups in order to create a better model.


In [7]:
# Use this function to check the performance of your model
rmse <- function(pred.label, truth.label){
    # Lower is better
    return(sqrt(mean((pred.label - truth.label)^2)))
}

In [8]:
# Input: model
# Print: RSS
rss <- function(fin.mod){
    pred.label <- predict(fin.mod, train[,-10])
    truth.label <- train[,10]
    print(rmse(pred.label, truth.label))
}

In [9]:
rss(sw.fit)

[1] 1.59735


Then, I create another function named rss.   
This function uses the model that we build as the input, call the rmse function to calculate the RSS of that model, and print out the result.  
In this way, we can increase the code re-use to make it clear.

In [10]:
#for ( a in seq(0, 10, by = 0.1)){
#    for ( b in seq(0, 10, by = 0.1)){
#        for ( c in seq(0, 10, by = 0.1)){
#            for ( d in seq(0, 10, by = 0.1)){
#    train <- read.csv("RegressionTrain.csv")
#    training <- train[1:1000,]
#    testing <-  train[1001:1400,]
#                
#    training$Fuel.Type[which(training$Fuel.Type == "G")] = a
#    training$Fuel.Type[which(training$Fuel.Type == "GM")] = b
#    training$Fuel.Type[which(training$Fuel.Type == "GP")] = c
#    training$Fuel.Type[which(training$Fuel.Type == "GPR")] = d
#    training$Fuel.Type <- as.numeric(training$Fuel.Type)
#    print(paste(a,b,c,d))
#    testing$Fuel.Type[which(testing$Fuel.Type == "G")] = a
#    testing$Fuel.Type[which(testing$Fuel.Type == "GM")] = b
#    testing$Fuel.Type[which(testing$Fuel.Type == "GP")] = c
#    testing$Fuel.Type[which(testing$Fuel.Type == "GPR")] = d
#    testing$Fuel.Type <- as.numeric(testing$Fuel.Type)
                
#    l.f <- lm(formula = Comb.FE~., training)
#    rss(l.f)
#}}}}

We find many characters attributes in the data, which is not convenient for us to build a model. As a result, we need to translate these character values into numeric values.  
In this part, we use specific numeric values to present each character's values.  
Here is a problem, which number is the best one to present a character value?   
To solve this, I create a for loop (you can see it in comment) from 0 to 10, step is 0.1, to try the different number for each variable, and create a multiple regression model using all the data. Then, calculate the RSS of this model. We use the numeric value which makes the RSS is minimum to present the character value.


In [11]:
# Use specific numeric values to present Lockup.Torque.Converter
train$Lockup.Torque.Converter[which(train$Lockup.Torque.Converter == "Y")] = 1
train$Lockup.Torque.Converter[which(train$Lockup.Torque.Converter == "N" )] = 2
train$Lockup.Torque.Converter <- as.numeric(train$Lockup.Torque.Converter)
# Use specific numeric values to present Drive.Sys
train$Drive.Sys[which(train$Drive.Sys == "P")] = 0.9
train$Drive.Sys[which(train$Drive.Sys == "4" )] = 3.8
train$Drive.Sys[which(train$Drive.Sys == "R")] = 5.2
train$Drive.Sys[which(train$Drive.Sys == "A")] = 2.9
train$Drive.Sys[which(train$Drive.Sys == "F")] = 17.3
train$Drive.Sys <- as.numeric(train$Drive.Sys)
# Use specific numeric values to present Aspiration
train$Aspiration[which(train$Aspiration == "SC")] = 3.7
train$Aspiration[which(train$Aspiration == "N" )] = 0.1
train$Aspiration[which(train$Aspiration == "OT")] = 1.4
train$Aspiration[which(train$Aspiration == "TC")] = 4.2
train$Aspiration[which(train$Aspiration == "TS")] = 4.1
train$Aspiration <- as.numeric(train$Aspiration)
# Use specific numeric values to present Fuel.Type
train$Fuel.Type[which(train$Fuel.Type == "G")] = 0.1
train$Fuel.Type[which(train$Fuel.Type == "GM")] = 0.2
train$Fuel.Type[which(train$Fuel.Type == "GP")] = 40
train$Fuel.Type[which(train$Fuel.Type == "GPR")] = 3
train$Fuel.Type <- as.numeric(train$Fuel.Type)

We choose a fancy model building method to create the model.     
To be specific, we fit a full model with logs of all variables, squares, and cubes of all variables. Also, it includes all interactions between predictors (two predictors multiplication or division, three predictors multiplication).  
Next step, we apply the stepwise selection procedure with the BIC penalty to prune out potentially less significant variables and get the final model.


In [12]:
b.fit <- lm(formula = Comb.FE~. + .*. + .*.*. + log(Drive.Sys)+ log(Eng.Displacement) 
             + log(Aspiration)+ log(No.Gears)+ log(Lockup.Torque.Converter)+ log(Fuel.Type)
             +log(Model.Year) + log(Max.Ethanol) + log(No.Cylinders)+ I(Eng.Displacement^2) 
             + I(No.Gears^2)+ I(Aspiration^2) + I(Lockup.Torque.Converter^2)+ I(Drive.Sys^2) 
             + I(Fuel.Type^2)+ I(Model.Year^2) + I(Max.Ethanol^2)+ I(No.Cylinders^2)  + I(No.Gears^3)+ I(Aspiration^3) + I(Lockup.Torque.Converter^3)+ I(Drive.Sys^3) 
             + I(Fuel.Type^3)+ I(Model.Year^3) + I(Max.Ethanol^3)+ I(No.Cylinders^3)
             + I(Drive.Sys/Eng.Displacement) + I(Drive.Sys/Aspiration) + I(Drive.Sys/No.Gears) 
             + I(Drive.Sys/Lockup.Torque.Converter) + I(Drive.Sys/Fuel.Type)          
             + I(Drive.Sys/Model.Year) + I(Drive.Sys/Max.Ethanol) + I(Drive.Sys/No.Cylinders) 
            + I(Eng.Displacement/Drive.Sys) + I(Eng.Displacement/Aspiration) + I(Eng.Displacement/No.Gears) 
            + I(Eng.Displacement/Lockup.Torque.Converter) + I(Eng.Displacement/Fuel.Type) + I(Eng.Displacement/Model.Year)
            + I(Eng.Displacement/Max.Ethanol) + I(Eng.Displacement/No.Cylinders)
            + I(Aspiration/Drive.Sys) + I(Aspiration/Eng.Displacement) + I(Aspiration/No.Gears) 
            + I(Aspiration/Lockup.Torque.Converter) + I(Aspiration/Fuel.Type) + I(Aspiration/Model.Year)
            + I(Aspiration/Max.Ethanol) + I(Aspiration/No.Cylinders)   
            + I(No.Gears/Drive.Sys) + I(No.Gears/Eng.Displacement) + I(No.Gears/Aspiration) 
            + I(No.Gears/Lockup.Torque.Converter) + I(No.Gears/Fuel.Type) + I(No.Gears/Model.Year)
            + I(No.Gears/Max.Ethanol) + I(No.Gears/No.Cylinders)
            + I(Lockup.Torque.Converter/Drive.Sys) + I(Lockup.Torque.Converter/Eng.Displacement) + I(Lockup.Torque.Converter/Aspiration) 
            + I(Lockup.Torque.Converter/No.Gears) + I(Lockup.Torque.Converter/Fuel.Type) + I(Lockup.Torque.Converter/Model.Year)
            + I(Lockup.Torque.Converter/Max.Ethanol) + I(Lockup.Torque.Converter/No.Cylinders)
            + I(Fuel.Type/Drive.Sys) + I(Fuel.Type/Eng.Displacement) + I(Fuel.Type/Aspiration) 
            + I(Fuel.Type/Lockup.Torque.Converter) + I(Fuel.Type/Lockup.Torque.Converter) + I(Fuel.Type/Model.Year)
            + I(Fuel.Type/Max.Ethanol) + I(Fuel.Type/No.Cylinders)
            + I(Model.Year/Drive.Sys) + I(Model.Year/Eng.Displacement) + I(Model.Year/Aspiration) 
            + I(Model.Year/Lockup.Torque.Converter) + I(Model.Year/Lockup.Torque.Converter) + I(Model.Year/Fuel.Type)
            + I(Model.Year/Max.Ethanol) + I(Model.Year/No.Cylinders)
            + I(Max.Ethanol/Drive.Sys) + I(Max.Ethanol/Eng.Displacement) + I(Max.Ethanol/Aspiration) 
            + I(Max.Ethanol/Lockup.Torque.Converter) + I(Max.Ethanol/Lockup.Torque.Converter) + I(Max.Ethanol/Fuel.Type)
            + I(Max.Ethanol/Model.Year) + I(Max.Ethanol/No.Cylinders)
            + I(No.Cylinders/Drive.Sys) + I(No.Cylinders/Eng.Displacement) + I(No.Cylinders/Aspiration) 
            + I(No.Cylinders/Lockup.Torque.Converter) + I(No.Cylinders/Lockup.Torque.Converter) + I(No.Cylinders/Fuel.Type)
            + I(No.Cylinders/Model.Year) + I(No.Cylinders/Max.Ethanol)
             , data = train) 

# Apply the stepwise selection procedure with the BIC penalty
back.fit = step(b.fit ,trace=0, k = log(nrow(train)), direction="both")

After the BIC procedure, a multiple regression model was built.   
We use this model to calculate the testing data value, and the RSS is 1.0289, which is smaller than the previous BIC model in Question 3 (RSS=1.5974). This means sw.fit model is much better than the previous one.   
This is my final model.

In [13]:
# Build your final model here, use additional coding block if you need
fin.mod <- back.fit

In order to use my model to predict the efficiency of the fuel, we need to do some preprocessing to the test data.  
To be specific, we need to change the character's values into some specific numeric values.  
After that, we can use my model to predict the values in the RegressionTest.csv document.


In [14]:
# Data wrangling the test data, in order to use the model.
# Use specific numeric values to present Fuel.Type
test$Fuel.Type[which(test$Fuel.Type == "G")] = 0.1
test$Fuel.Type[which(test$Fuel.Type == "GM")] = 0.2
test$Fuel.Type[which(test$Fuel.Type == "GP")] = 40
test$Fuel.Type[which(test$Fuel.Type == "GPR")] = 3
test$Fuel.Type <- as.numeric(test$Fuel.Type)
# Use specific numeric values to present Aspiration
test$Aspiration[which(test$Aspiration == "SC")] = 3.7
test$Aspiration[which(test$Aspiration == "N" )] = 0.1
test$Aspiration[which(test$Aspiration == "OT")] = 1.4
test$Aspiration[which(test$Aspiration == "TC")] = 4.2
test$Aspiration[which(test$Aspiration == "TS")] = 4.1
test$Aspiration <- as.numeric(test$Aspiration)
# Use specific numeric values to present Lockup.Torque.Converter
test$Lockup.Torque.Converter[which(test$Lockup.Torque.Converter == "Y")] = 1
test$Lockup.Torque.Converter[which(test$Lockup.Torque.Converter == "N" )] = 2
test$Lockup.Torque.Converter <- as.numeric(test$Lockup.Torque.Converter)
# Use specific numeric values to present Drive.Sys
test$Drive.Sys[which(test$Drive.Sys == "P")] = 0.9
test$Drive.Sys[which(test$Drive.Sys == "4" )] = 3.8
test$Drive.Sys[which(test$Drive.Sys == "R")] = 5.2
test$Drive.Sys[which(test$Drive.Sys == "A")] = 2.9
test$Drive.Sys[which(test$Drive.Sys == "F")] = 17.3
test$Drive.Sys <- as.numeric(test$Drive.Sys)

In [15]:
# If you are using any packages that perform the prediction differently, please change the value of this variable
pred.label <- predict(fin.mod, test)

Warning message in predict.lm(fin.mod, test):
"prediction from a rank-deficient fit may be misleading"


In [16]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# put this label in a csv file to commit to the Leaderboard
write.csv(data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
          "RegressionPredictLabel.csv", row.names = F)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
## Please skip (don't run) this if you are a student
## For teaching team use only
RMSE.fin <- rmse(pred.label, label$Label)
cat(paste("RMSE is", RMSE.fin))

<h1>Part 2: Classification (50 Marks)</h1>

In this part, you are going to work with "Census Income Dataset" which was originally donated by Ronny Kohavi and Barry Becker to UCI (University of California, Irvine) in 1996. This is a trimmed dataset used for machine learning students to study classification. 

This dataset has collected over 40,000 records (we excluded some data in our version) regarding personal yearly income with 12 attributes (predictors). The attributes comprise many aspects of a person that may contribute to the yearly income. You can use summary() function to obtain the attributes information. Your prediction task is to determine whether a person makes over 50K a year.

We have splitted the dataset into a trainning and a testing set. There are 27245 records in the training set while 13631 records in the testing set. Besides the 12 predictors, there is one more column named Salary indicating whether a person's yearly income is over 50K. The label information is a seperated file for the testing set and will be used by us to asess your performance later. Note the label TRUE means an individual's yearly salary exceeds 50K while FALSE means an individual's yearly salary is under 50K.

$\textbf{Note:}$ If not explicitly mentioned, libraries are not allowed

In [17]:
# Read the data from students' side
remove(list = ls())
train <- read.csv("ClassTrain.csv")
test  <- read.csv("ClassTest.csv")

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
# Please skip (don't run) this if you are a student
# Read in the data from marking tutors' side (ensure no cheating!)
remove(list = ls())
train <- read.csv("../data/ClassTrain.csv")
test  <- read.csv("../data/ClassTest.csv")
label <- read.csv("../data/ClassTestLabel.csv")

<h2> Question 1 (10 Marks) </h2>

Fit a $\textbf{Generalized Linear Model (Logistic Regression)}$ to predict level of income (salary) $\left(\;\geq 50\;\text{K, or } <50\;\text{K}\;\right)$ using the ``train`` dataset. Using the results of fitting this model, which predictors do you think are possibly associated with the level of Salary (use ``0.05`` significant level), and why? Which ``three variables`` appear to be the strongest predictors of salary, and why? 

Furthermore, you can see that you have much more predictors in this part than in the ``linear model`` from Part 1 $\Rightarrow$ manually checking information is counterproductive. Thus, please write a function to automate these processes $\textbf{(1)}$ selecting important feature against 0.05 threshold and $\textbf{(2)}$ Selecting three most important features.

$\textbf{Note}$: You don't have to worry about categorical variables here since R can deal with this automatically, focus your efforts on interpretation. Additionally, when explaining why features are strongly associated with the target, please refrain from giving one or two sentences answers, these answers are not descriptive and will result in a deduction of marks. Finally, please name the model here ``glm.fit`` and have the parameter in the model set to ``family = binomial``.

$\textbf{YOUR ANSWER HERE}$  
After building a generalized linear model (logistic regression), we can use a summary function to see the result of this model.
###### Selecting important feature
We find that Age, FinalWeight,WorkClassLocal-gov, WorkClassPrivate, WorkClassSelf-emp-inc, WorkClassSelf-emp-not-inc, Education7th-8th, EducationAssoc-acdm,EducationAssoc-voc, EducationBachelors,EducationDoctorate, EducationHS-grad, EducationMasters, EducationProf-school, EducationSome-college, MaritalStatus, Married-civ-spouse,  WorkClassState-gov, MaritalStatusNever-married, OccupationExec-managerial, OccupationFarming-fishing,  OccupationHandlers-cleaners, OccupationMachine-op-inspct, OccupationOther-service, OccupationProf-specialty, OccupationProtective-serv, OccupationSales, OccupationTech-support,RelationshipNot-in-family, RelationshipOwn-child, RelationshipWife, RaceAsian-Pac-Islander, RaceWhite, CapitalGain, CapitalLoss, WorkClass, GenderMale, and HoursWork are possibly associated with the level of Salary.  
From the summary result, the significant level of these values are smaller than 0.05. So we can conclude that these predictors are possibly associated with the level of Salary. 
###### Selecting three most important features
In order to choose the three strongest variables predictors of salary, we need to find the parameters with the smallest values of significance. (*** in the coefficients of the summary result). We would focus on the numeric values which consist of more information.  
So, we have Age, CapitalGain, CapitalLoss, FinalWeight, and HoursWork.   
Then, we need to focus on the raw data to have a general view of these parameters. Unfortunately, CapitalGain and CapitalLoss column contains many 0 values, which is useless to predict. So, we need to discard them.  
In the end, we find out three variables appear to be the strongest predictors of salary.   
They are Age, FinalWeight and HoursWork.



In [18]:
# Build your model, keep family = binomial, ignore the warnings, they are benign
glm.fit <- glm(Salary ~ ., data=train, family=binomial)
summary(glm.fit)

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"



Call:
glm(formula = Salary ~ ., family = binomial, data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-5.1013  -0.5296  -0.1926   0.0276   3.4349  

Coefficients:
                                     Estimate Std. Error z value Pr(>|z|)    
(Intercept)                        -7.614e+00  4.525e-01 -16.826  < 2e-16 ***
Age                                 2.626e-02  1.779e-03  14.762  < 2e-16 ***
WorkClassLocal-gov                 -7.214e-01  1.168e-01  -6.179 6.46e-10 ***
WorkClassPrivate                   -4.734e-01  9.693e-02  -4.884 1.04e-06 ***
WorkClassSelf-emp-inc              -2.974e-01  1.283e-01  -2.317 0.020506 *  
WorkClassSelf-emp-not-inc          -9.994e-01  1.139e-01  -8.772  < 2e-16 ***
WorkClassState-gov                 -7.757e-01  1.294e-01  -5.996 2.03e-09 ***
FinalWeight                         7.896e-07  1.822e-07   4.334 1.46e-05 ***
Education11th                       6.909e-02  2.201e-01   0.314 0.753589    
Education12th            

<h2> Question 2 (10 Marks) </h2>

Firstly, please use the model created in the previous question to predict for the labels of the $\textbf{train}$ data. Consequently, our objective is to compare this ``predict.label`` with the ``truth.label`` from the $\textbf{test}$ data. However, as we don't know the $\textbf{test}$ label, we have to estimate model performance using $\textbf{train}$ data at this moment.

Secondly, since our objective is to estimate the performance of this model in making correct predictions; thus, this question also asks you to explore different [performance metrics](https://en.wikipedia.org/wiki/Precision_and_recall) for classification models. The metrics we will use are $\textbf{Accuracy, Precision, Recall, and F1 Score}$, please create a function to calculate these value and print them out properly using the given structure.

Additionally, please also discuss the results of these values in the context of your model.

$\textbf{Note}$: This asks for your descriptions, please refrain from using one or two lines to describe/discuss the effect. Keep answers to be 4 decimal places

In [19]:
# Apply your previous model to perform prediction, keep type = "response"
# Don't worry if you receive some warnings, they are benign
predict.label <- predict(glm.fit, train)
# Truth label from train data
truth.label <- train$Salary

In [20]:
# Model statistics function
mod.stat <- function(predict.label, truth.label){
    # instantiate the variables
    accuracy <- NULL
    precision <- NULL
    recall <- NULL
    F1 <- NULL
    
    tp <- 0
    fp <- 0
    fn <- 0 
    tn <- 0
    ##############################
    #Your calculatation here
    for (i in seq(1, length(truth.label))){
    # Change the boolean to character in truth.label
    if (truth.label[i] == "TRUE"){
        truth.label[i] = "1"
    }   
    if (truth.label[i] == "FALSE"){
        truth.label[i] = "0"
    }   
    # Change the numeric to character in predict.label
    if (predict.label[i] >=0.5){
        predict.label[i] = "1"
    }
    if (predict.label[i] <0.5){
        predict.label[i] = "0"
    }  
    
    # If the answer is right
    if (predict.label[i] == truth.label[i]){
        # True positive
        if (predict.label[i] == "1"){
            tp = tp + 1
        }
        # True negative
        if (predict.label[i] == "0"){
            tn = tn + 1
        }
    }
    
    # If the answer is wrong
    if (predict.label[i] != truth.label[i]){
        # False positive
        if (predict.label[i] == "1"){
            fp = fp + 1
        }
        # False negative
        if (predict.label[i] == "0"){
            fn = fn + 1
        }
    }
}
     # Keep answers to be 4 decimal places
     accuracy = round( ( tp + tn ) / (tp + tn + fp + fn) , 4 )
     precision = round( tp  / (tp + fp) , 4 )
     recall = round( tp  / (tp + fn) , 4 )
     F1 = round( 2*tp / (2*tp + fp +fn) , 4 )
    ##############################
    
    # Return a list of value
    return(list("accuracy" = accuracy, "precision" = precision, "recall" = recall, "fscore" = F1))
}

In [21]:
# Run the function to get statistics, provide description/discussion after this
mod.stat(predict.label, truth.label)

$accuracy
[1] 0.8386

$precision
[1] 0.8052

$recall
[1] 0.4879

$fscore
[1] 0.6076

$\textbf{YOUR ANSWER HERE}$  
From the results of my model, we find that the accuracy and precision are pretty high, but the recall is lower than 50%. 
This means this model is poor in figuring out True positive and good at True negative.   
To be specific, it misclassifies many true values to false. On the other hand, it performs well at recognizing false value.


<h2> Question 3 (5 Marks) </h2>

Use the stepwise selection procedure with the $\textbf{BIC}$ penalty to prune out potentially unimportant variables. Checking the performance of your model using the created ``mod.stat()`` function, please give your discussion as how this model is compared with the ``glm.fit``(you can run the ``mod.stat()`` function for this as well if you want to).

$\textbf{Note}$: please don't change the default direction ``both`` in the step function, this is so that we can check your work easily. Additionally, please name this model ``sw.fit``. Don't worry about the warnings, they are benign

$\textbf{YOUR ANSWER HERE}$

In [22]:
# Setting to suppress warnings
options(warn=-1)
# Fit a stepwise model
sw.fit <- step(glm.fit, trace=0, k = log(length(truth.label)), direction="both")
# Setting to suppress warnings
options(warn=0)
# Getting the summary to understand the result
summary(sw.fit)


Call:
glm(formula = Salary ~ Age + WorkClass + FinalWeight + Education + 
    MaritalStatus + Occupation + Relationship + Gender + CapitalGain + 
    CapitalLoss + HoursWork, family = binomial, data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-5.0961  -0.5291  -0.1936   0.0279   3.4393  

Coefficients:
                                     Estimate Std. Error z value Pr(>|z|)    
(Intercept)                        -7.130e+00  3.929e-01 -18.146  < 2e-16 ***
Age                                 2.644e-02  1.778e-03  14.869  < 2e-16 ***
WorkClassLocal-gov                 -7.159e-01  1.163e-01  -6.154 7.54e-10 ***
WorkClassPrivate                   -4.588e-01  9.626e-02  -4.766 1.88e-06 ***
WorkClassSelf-emp-inc              -2.776e-01  1.278e-01  -2.173  0.02977 *  
WorkClassSelf-emp-not-inc          -9.857e-01  1.133e-01  -8.703  < 2e-16 ***
WorkClassState-gov                 -7.653e-01  1.290e-01  -5.930 3.02e-09 ***
FinalWeight                         7.

In [23]:
# Making prediction using train data and view the statistics
predict.label.sw <- predict(sw.fit, train)
# Only run the below if you have labels, in your submission, this must be UNCOMMENTED
mod.stat(predict.label.sw, truth.label)

$accuracy
[1] 0.8387

$precision
[1] 0.8062

$recall
[1] 0.4876

$fscore
[1] 0.6077

PROVIDE DISCUSSION HERE  

In this model, it has removed many variables, such as Race, so the resulting model is quite a bit simpler.  
But when we use mod.stat function to check the performance metrics for classification models, we find that the improvement is very limited. To be specific, the accuracy and precision increase by 0.0001 and 0.001 respectively.  
In summary, using the stepwise selection procedure with the BIC penalty is useless in this question, it improves the model performs slightly.


<h2> Question 4 (Libraries are allowed) (25 Marks) </h2>

Similar to the first part, to simulate for a realistic modelling process, this question will be in the form of a competition among students to find out who has the best model.

Thus, You will be graded by the performance of your model compared to your classmates', the better your model, the higher your score. Additionally, you need to write a short paragraph describing/documenting your thought process in this model building process ``(300 words)``. Note that this is to explain to us why you build your current model so that we can verify that you understand the model you build and not just copy from other people.

$\textbf{Note}$ Please make sure that we can install the libraries that you use in this part, the code structure can be:

``install.packages("some package", repos='http://cran.us.r-project.org')``

``library("some package")``

Remember that if we cannot run your code, we will have to give you a deduction, our suggestion is for you to use the standard ``R version 3.6.1``

You also need to name your final model ``fin.mod`` so we can run a check to find out your performance. A good test for your understanding would be to set the previous $\textbf{BIC model}$ to be the final model to check if your code works perfectly.


$20$ Marks for the model performance in the competition

$5$ Marks for logically writing down the thought process in building the final model

This is the [link](https://www.kaggle.com/t/1bdebc96607742dbaf47ab36cd3ae421) to the competition

$\textbf{YOUR ANSWER HERE}$  
I write a short paragraph describing/documenting my thought process in this model building process under each code block. 

In [24]:
#for ( a in seq(0, 10, by = 0.1)){
#train <- read.csv("ClassTrain.csv")
#train$WorkClass[which(train$WorkClass == "Federal-gov" )] = a
#train$WorkClass[which(train$WorkClass == "Private" )] = 0
#train$WorkClass[which(train$WorkClass == "Self-emp-inc" )] = 0
#train$WorkClass[which(train$WorkClass == "Self-emp-not-inc" )] = 0
#train$WorkClass[which(train$WorkClass == "State-gov" )] = 0
#train$WorkClass[which(train$WorkClass == "Local-gov" )] = 0
#train$WorkClass <- as.numeric(train$WorkClass)

#glm.fit <- glm(Salary ~ ., data=train, family=binomial)
#predict <- predict(glm.fit, train)
#mod.stat(predict, truth.label)
#}

We find many characters attributes in the data, which is not convenient for us to build a model. As a result, we need to translate these character values into numeric values.  
In this part, we use specific numeric values to present each character's values.  
Here is a problem, which number is the best one to present a character value?   
To solve this, I create a for loop (you can see it in comment) from 0 to 10, step is 0.1, to try the different number for each variable, and create a multiple regression model using all the data.  
Then, we will call the mod.stat function to calculate the performance metrics. We will use the specific number, to get the lowest accuracy, precision, and recall to present that character.


In [25]:
# Use specific numeric values to present WorkClass
train$WorkClass[which(train$WorkClass == "Federal-gov" )] = 0.01
train$WorkClass[which(train$WorkClass == "Private" )] = 0.5
train$WorkClass[which(train$WorkClass == "Self-emp-inc" )] = 2.6
train$WorkClass[which(train$WorkClass == "Self-emp-not-inc" )] = 4.4
train$WorkClass[which(train$WorkClass == "State-gov" )] = 3.8
train$WorkClass[which(train$WorkClass == "Local-gov" )] = 6
train$WorkClass <- as.numeric(train$WorkClass)
# Use specific numeric values to present Race
train$Race[which(train$Race == "Amer-Indian-Eskimo" )] = 1.2
train$Race[which(train$Race == "Asian-Pac-Islander" )] = 5
train$Race[which(train$Race == "Black" )] = 0.01
train$Race[which(train$Race == "Other" )] = 0.01
train$Race[which(train$Race == "White" )] = 5
train$Race <- as.numeric(train$Race)
# Use specific numeric values to present Gender
train$Gender[which(train$Gender == "Female" )] = 1
train$Gender[which(train$Gender == "Male" )] = 2
train$Gender <- as.numeric(train$Gender)
# Use specific numeric values to present MaritalStatus
train$MaritalStatus[which(train$MaritalStatus == "Divorced" )] = 1
train$MaritalStatus[which(train$MaritalStatus == "Married-civ-spouse" )] = 3
train$MaritalStatus[which(train$MaritalStatus == "Married-spouse-absent" )] = 0.01
train$MaritalStatus[which(train$MaritalStatus == "Never-married" )] = 0.01
train$MaritalStatus[which(train$MaritalStatus == "Separated" )] = 0.01
train$MaritalStatus[which(train$MaritalStatus == "Widowed" )] = 0.01
train$MaritalStatus <- as.numeric(train$MaritalStatus)

We also choose a fancy model building method to create the model.      
To be specific, we fit a full model with logs of all variables, and squares of all variables.  
Next step, we apply the stepwise selection procedure with the BIC penalty to prune out potentially less significant variables and get the final model.


In [26]:
options(warn=-1)
glm.fit <- glm(Salary ~ .+  log(WorkClass)+ log(Race)+ log(MaritalStatus)
             +log(HoursWork) + log(Age) + log(FinalWeight)+ I(WorkClass^2) 
             + I(Race^2)+ I(MaritalStatus^2) + I(HoursWork^2)+ I(FinalWeight^2) 
             + I(Age^2), data=train, family=binomial)
sw.fit <- step(glm.fit,trace=0, k = log(length(train[,1])), direction="both")
predict.label.sw <- predict(sw.fit, train)
mod.stat(predict.label.sw, truth.label)

$accuracy
[1] 0.8432

$precision
[1] 0.8159

$recall
[1] 0.5009

$fscore
[1] 0.6207

After the BIC procedure, we get our final model.  
We use this model to calculate the different performance metrics for classification models. We get 0.8432 of accuracy and 0.8159 of precision, which is better than the previous model in question 3.


In [27]:
options(warn=-1)
# Build your final model here, use additional coding block if you want to
glm.fit <- glm(formula = Salary ~ Age + Education + FinalWeight + MaritalStatus + Occupation + 
    Relationship + Race + Gender + CapitalGain + CapitalLoss + 
    HoursWork + log(WorkClass) + log(Age) + log(HoursWork) +
    log(FinalWeight), family = binomial, data = train)

Next part is about predict the test data.  
In order to use my model to predict the level of salary, we need to do some preprocessing to the raw data.  
After that, we can use my model to predict the test data.

In [28]:
# Use specific numeric values to present WorkClass
test$WorkClass[which(test$WorkClass == "Federal-gov" )] = 0.01
test$WorkClass[which(test$WorkClass == "Private" )] = 0.5
test$WorkClass[which(test$WorkClass == "Self-emp-inc" )] = 2.6
test$WorkClass[which(test$WorkClass == "Self-emp-not-inc" )] = 4.4
test$WorkClass[which(test$WorkClass == "State-gov" )] = 3.8
test$WorkClass[which(test$WorkClass == "Local-gov" )] = 6
test$WorkClass <- as.numeric(test$WorkClass)
# Use specific numeric values to present Race
test$Race[which(test$Race == "Amer-Indian-Eskimo" )] = 1.2
test$Race[which(test$Race == "Asian-Pac-Islander" )] = 5
test$Race[which(test$Race == "Black" )] = 0.01
test$Race[which(test$Race == "Other" )] = 0.01
test$Race[which(test$Race == "White" )] = 5
test$Race <- as.numeric(test$Race)
# Use specific numeric values to present Gender
test$Gender[which(test$Gender == "Female" )] = 1
test$Gender[which(test$Gender == "Male" )] = 2
test$Gender <- as.numeric(test$Gender)
# Use specific numeric values to present MaritalStatus
test$MaritalStatus[which(test$MaritalStatus == "Divorced" )] = 1
test$MaritalStatus[which(test$MaritalStatus == "Married-civ-spouse" )] = 3
test$MaritalStatus[which(test$MaritalStatus == "Married-spouse-absent" )] = 0.01
test$MaritalStatus[which(test$MaritalStatus == "Never-married" )] = 0.01
test$MaritalStatus[which(test$MaritalStatus == "Separated" )] = 0.01
test$MaritalStatus[which(test$MaritalStatus == "Widowed" )] = 0.01
test$MaritalStatus <- as.numeric(test$MaritalStatus)

After that, we can get the result.  
However, our result is numeric, we need to use some method to change the numeric format to logical values, which consist of "TRUE" or "FALSE" value. We set 0.5 as a threshold value, if the value larger than 0.5, we think they are TRUE. Otherwise, we regard them as FALSE.


In [29]:
# Getting the predict label for the TEST data
pred.label <- predict(glm.fit, test)
# Change the numeric number to logical
for (i in seq(1:length(pred.label))){
    if (pred.label[i] >=0.5){
        pred.label[i] = "TRUE"
    }
    if (pred.label[i] <0.5){
        pred.label[i] = "FALSE"
    }  
}
pred.label <- as.logical(pred.label)

In [30]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# Use this csv file to commit to the leaderboard
write.csv(data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
          "ClassPredictLabel.csv", row.names = F)

In [31]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
## Please skip (don't run) this if you are a student
## For teaching team use only
source("../data/modassess.r")
model.perf <- mod.stat.test(pred.label,label$Label)
print(model.perf)

ERROR: Error in file(filename, "r", encoding = encoding): cannot open the connection


<h1>References</h1>